cache_flat_mutation_reader.hh

```diff
@@ -730,9 +749,44 @@ void cache_flat_mutation_reader::copy_from_cache_to_buffer() {
         }
     }
 
-    // We add the row to the buffer even when it's full.
-    // This simplifies the code. For more info see #3139.
     if (_next_row_in_range) {
+        bool remove_row = false;
+
+        if (_read_context.tombstone_gc_state() // do not compact rows when tombstone_gc_state is not set (used in some unit tests)
+            && !_next_row.dummy()
+            && _snp->at_latest_version()
+            && _snp->at_oldest_version()) {
+            deletable_row& row = _next_row.latest_row();
+            auto t = row.deleted_at();
+
+            auto row_tomb_expired = [&](row_tombstone tomb) {
+                return (tomb && tomb.max_deletion_time() < _gc_before);
+            };
+
+            auto is_row_dead = [&](const deletable_row& row) {
+                return (!row.marker().is_missing() && row.marker().is_dead(_read_time));
+            };
+
+            if (row_tomb_expired(t) || is_row_dead(row)) {
+                can_gc_fn always_gc = [&](tombstone) { return true; };
+                row_marker marker = row.marker();
+
+                _read_context.cache()._tracker.on_row_compaction_attempt();
+
+                try {
+                    with_allocator(_snp->region().allocator(), [&] {
+                        deletable_row row_copy(*_schema, row);
```

Comment:
After #13761 (which made schema upgrades preemptible), `latest_row()`'s schema can be older than the reader's. So you shouldn't use `_schema` here, but the row's schema. It's stored in `_current_row[0].schema` in the cursor.

Reply:
Aw, thank you. Will make changes now

Reply:
> If I understand right we should change _schema to _current_row[0].schema in every call connected to row?

No, only in the calls which reach directly to `latest_row()`.

The data structure under partition_snapshot is a chain of partition_versions, each of which is a b-tree of clustering rows. New versions are added when the cache is updated (this happens when a memtable is flushed), or when the schema is upgraded. (And are later merged with the older versions by a background worker, when snapshot isolation is no longer needed).

Since we recently reused this mechanism (https://github.com/scylladb/scylladb/pull/13761) to make schema upgrades in cache entries preemptible, versions in the chain can have different schemas, and each row entry should be interpreted with the schema of its containing version.

Normally it's the cursor job to deal with versions, but when we call `latest_row()`, we are reaching under the cursor directly to a particular row entry, and we have to take care to use the correct schema manually.   

For example, the state can look like that:
```
                      v3,s2     v2,s1     v1,s1 

...
ck=3                                       e1_2
ck=2 (_next_row)                 e2_2      e2_1
ck=1                  e1_1                  
...
```
Here, the snapshot covers 3 versions: v1, v2, v3 (from oldest to newest).
v1 and v2 have schema s1, while v3 has schema s2.
The cursor points at clustering key `2`, which has entries in versions v1 and v2, but not v3.
The schema of the read/snapshot is s2, but latest_row() points at e2_2, which has s1. To interpret e2_2 properly, we have to pass s1 to functions which access it.

Reply:
I see you changed it in one more place (`            _lower_bound = position_in_partition::after_key(_next_row.latest_row_schema(), _next_row.position());`), and that's somewhat inappropriate (because we dealing with the reader's position, not the latest row's position), but it's okay because keys/positions can't change during schema upgrades. (@tgrabiec please confirm).

Reply:
Thank you for such a detailed explanation! I've got the idea. You're right: it's more correct to use `latests_row_schema()` for `latest_row() `only. I will fix it.